userspace/libsinsp/container_engine/docker.h

```diff
@@ -43,20 +51,64 @@ class docker
 	};
 
 public:
-	docker();
+	docker_async_source(uint64_t max_wait_ms, uint64_t ttl_ms);
+	virtual ~docker_async_source();
 
-	bool resolve(sinsp_container_manager* manager, sinsp_threadinfo* tinfo, bool query_os_for_missing_info);
-	static void cleanup();
+	void set_inspector(sinsp *inspector);
 	static void set_query_image_info(bool query_image_info);
-	static void parse_json_mounts(const Json::Value &mnt_obj, std::vector<sinsp_container_info::container_mount_info> &mounts);
+
+	// Note that this tid is the current top tid for this container
+	void set_top_tid(const std::string &container_id, sinsp_threadinfo *tinfo);
+
+	// Update the mapping from container id to top running tid for
+	// that container.
+	void update_top_tid(std::string &container_id, sinsp_threadinfo *tinfo);
+
+	// Get the thread id of the top thread running in this container.
+	int64_t get_top_tid(const std::string &container_id);
+
+        bool pending_lookup(std::string &container_id);
 
 protected:
-	docker_response get_docker(sinsp_container_manager* manager, const std::string& url, std::string &json);
+	void run_impl();
+
```

Comment:
I don't think so--the *top_tid* and pending_lookup() methods are all called from the docker object, set_query_image_info is called from the container manager.

Reply:
Sorry -- I meant the members that are currently `protected`.

    ...
    protected:
             void run_impl();

    private:
            std::string build_request(const std::string& url);
            docker_response get_docker(const std::string& url, std::string &json);
            ...

Reply:
Oh sure, I can make them private. My habit is protected by default just in case you want to derive from it later, but that's very unlikely for this class.